### ISSUE TITLE
Salt integration questions

### ISSUE THREAD

**Issue by powellchristoph** [conversis/varstack#2]
Greetings, 

I was digging through the ext_pillar's in Saltstack to solve a problem and I stumbled across varstack. It is extremely interesting as it may solve my exact problem. Unfortunately, there doesn't appear to be any documentation for actually using varstack as an ext_pillar in Salt.

I see that you are passing in **grains** to varstack.evaluate(), but for what purpose? Also, how do you pass the '%{variable}' into varstack from the ext_pillar to access the various stacks?

My use case is simple, but so far difficult to solve. We have multiple datacenter locations and sometimes specific pillar data for them. I would like to be able to access the specific pillar data for a location if it exists, otherwise retrieve it from a set of defaults. Varstack appears to be perfect.

Thanks for any information!

**Comment by dennisjac** [comment]
Thanks for the interest in varstack!

Using varstack in salt is fairly simple. Just put the following into the config file of your master:

```
ext_pillar:
  - varstack: /etc/varstack.yaml
```

Varstack will then use /etc/varstack.yaml to determine which configuration data to return as pillar information. From there you can take a look at the README of varstack on how this file is evaluated.

The grains can be used in the configuration mentioned above to control which configuration files are used to build the pillar. Let's assume /etc/varstack.yaml contains the following:

```
datadir: /etc/varstack/stack
stack:
    - defaults
    - roles/%{role}
    - nodes/%{fqdn}
```

Also let's assume that the "role" grain isn't set at all and that the "fqdn" grain contains "webhost1.example.com". Given these assumptions varstack will do the following:
1. Read the file /etc/varstack/stack/default.yaml (based on the "datadir" and first entry under "stack")
2. Ignore the "roles/%{role}" entry because the "role" grain isn't defined (as per out assumptions above)
3. Try to read the file /etc/varstack/stack/nodes/webhost1.example.com.yaml and if it exists merge the data into the data already read so far (only the default.yaml in this case).
4. Return the final data set as pillar.

Now let's say on the host you created a grain "role" and set it to the value "webserver". The evaluation will then work as follows:
1. Read the file /etc/varstack/stack/default.yaml (based on the "datadir" and first entry under "stack")
2. Try to read the file /etc/varstack/stack/roles/webserver.yaml and if it exists merge the data into the data already read so far (only the default.yaml so far).
3. Try to read the file /etc/varstack/stack/nodes/webhost1.example.com.yaml and if it exists merge the data into the data already read so far (the merged content from default.yaml and webserver.yaml).
4. Return the final data set as pillar.

So basically you want to define your global defaults in the first file evaluated (here default.yaml) and then add additional levels that can override or add to that information. You could for example define a default name server in default.yaml and then have an entry "datacenter/%{datacenter}" in the above config and set the "datacenter" grain for all systems running there and in that file override the name server information so the systems use a local name server in that facility rather then the default one.

I hope this helps you with your setup and if you have additional questions don't hesitate to ask.

**Comment by bbinet** [comment]
Thanks @dennisjac for varstack and the saltstack integration doc available in this issue. I think it could help a few users if this was available directly in the salt documentation:
http://docs.saltstack.com/en/latest/ref/pillar/all/salt.pillar.varstack_pillar.html

**Comment by powellchristoph** [comment]
:+1:

**Comment by dennisjac** [comment]
That's a good point. I'll create a pull request for saltstack to include this in the documentation.

**Comment by dennisjac** [comment]
Done: https://github.com/saltstack/salt/pull/23691

**Comment by bbinet** [comment]
Thanks
